File: day-39-cf-797-c.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main()
{
    std::string s;

    std::cin >> s;

    int num_chars = (int)s.length();

    std::vector<char> a_stack;

    for(size_t i = s.length(); i != 0; i--) {
        a_stack.push_back(s[i - 1]);
    }

    /**
     * Store positions of all occurrences of each characters in the "a" stack counting from the bottom, in order from
     * top to bottom
     */

    std::vector<std::vector<int>> occurrences(26);

    for(int i = (int)a_stack.size() - 1; i >= 0; i--) {
        occurrences[a_stack[i] - 'a'].push_back(i);
    }

    std::vector<char> b_stack;
    std::vector<char> c_stack;

    char next_min_char;

    for(next_min_char = 'a'; occurrences[next_min_char - 'a'].empty(); next_min_char++) { }

    int next_occurrence_index = 0;

    while(!a_stack.empty()) {
        while(a_stack.size() > occurrences[next_min_char - 'a'][next_occurrence_index]) {
            b_stack.push_back(a_stack.back());

            a_stack.pop_back();
        }

        c_stack.push_back(b_stack.back());

        b_stack.pop_back();

        next_occurrence_index += 1;

        if(a_stack.empty()) {
            break;
        }

        while(next_occurrence_index == occurrences[next_min_char - 'a'].size()) {
            for(next_min_char += 1; occurrences[next_min_char - 'a'].empty(); next_min_char++) { }

            while(b_stack.back() == next_min_char) {
                c_stack.push_back(b_stack.back());

                b_stack.pop_back();
            }

            while(occurrences[next_min_char - 'a'][next_occurrence_index] >= a_stack.size()
                  && next_occurrence_index >= occurrences[next_min_char - 'a'].size()) {
                next_occurrence_index += 1;
            }
        }
    }

    while(!b_stack.empty()) {
        c_stack.push_back(b_stack.back());

        b_stack.pop_back();
    }

    for(char c : c_stack) {
        std::cout << c;
    }

    std::cout << std::endl;
}

File: day-39-cf-797-c.py

s = raw_input()

a_stack = []

a_stack_occurrences = {chr(c): 0 for c in xrange(ord('a'), ord('z') + 1)}

for c in reversed(s):
    a_stack.append(c)

    # Maintain the number of times each symbol occurs in the "a" stack (for faster minimum extraction)
    a_stack_occurrences[c] += 1

b_stack = []
c_stack = []

while len(a_stack) != 0:
    # It is always the best to extract the minimum character from the "a" stack and put it in front of the string

    min_char = filter(lambda k: a_stack_occurrences[k] != 0, a_stack_occurrences.keys())[0]

    # If some occurrences of the current `min_char` or better character accidentally got on top of the "b" stack while
    # the previous `min_char` was extracted, put them on top of the "c" stack not to bury them when extracting the
    # current `min_char`
    while len(b_stack) > 0 and b_stack[-1] <= min_char:
        c_stack.append(b_stack.pop())

    # Extract the `min_char`
    while a_stack[-1] != min_char:
        top = a_stack.pop()

        b_stack.append(top)

        a_stack_occurrences[top] -= 1

    # The actual legal transfer would be a -> b -> c
    top = a_stack.pop()

    c_stack.append(top)

    a_stack_occurrences[top] -= 1

# Dump all that's remaining on the "b" stack onto the "c" stack. Nothing else can be done at this point
while len(b_stack) != 0:
    c_stack.append(b_stack.pop())

print ''.join(c_stack)
